### PR TITLE
Module: Add badgeText prop

### PR DESCRIPTION
**Pull request by dangerismycat**
[Figma designs](https://www.figma.com/file/B5S5iyikCHqKUcmofquJOe/Web-Creation-Improvements?node-id=136%3A5387)

This PR adds an optional `badgeText` prop to Module (and related Module components) to display a badge after the title text.

Screen Shot 2021-04-28 at 12 53 36 PM

Screen Shot 2021-04-28 at 1 08 49 PM

## Test Plan
Check out the new "Static – Badge" and "Expandable - Icon and Badge " sections in the docs.

### CODE REVIEW

File: docs/src/Module.doc.js [hunk 1]
```diff
@@ -24,17 +24,17 @@ card(
...
+        href: 'static-badge',
         type: 'string',
-        required: true,
-        description: 'Id to identify this Module',
+        description:
+          'Add a badge displayed after the title. Will not be displayed if `title` is not provided. Be sure to localize the text.',
       },
       {
-        name: 'title',
+        name: 'children',
```
Comment (dangerismycat) [inline docs/src/Module.doc.js:34]: git is being weird, this is just moved around to alphabetize the props

File: docs/src/Module.doc.js [hunk 2]
```diff
@@ -24,17 +24,17 @@ card(
     id="static-Module"
     props={[
       {
-        name: 'id',
-        href: 'static-default',
+        name: 'badgeText',
```
Comment (dangerismycat) [inline docs/src/Module.doc.js:27]: this one is new

File: docs/src/Module.doc.js [hunk 3]
```diff
@@ -46,7 +46,21 @@ card(
         name: 'iconAccessibilityLabel',
         href: 'static-icon',
         type: 'string',
-        description: 'Label to provide information about the icon used for screen readers',
+        description:
+          'Label to provide information about the icon used for screen readers. Be sure to localize the label.',
```
Comment (dangerismycat) [inline docs/src/Module.doc.js:50]: These aren't new, just added notes about localization (and reworded the id description)

File: docs/src/Module.doc.js [hunk 4]
```diff
@@ -103,21 +104,36 @@ card(
...
+        required: true,
+        description: 'Unique id to identify this Module',
       },
       {
         name: 'items',
         href: 'expandable-items',
-        type:
-          'Array<{| title: string, icon?: $Keys<typeof icons>, summary?: Array<string>, type?: "info" | "error", iconAccessibilityLabel?: string, children: ?React.Node |}>',
+        type: `
+        Array<{|
```
Comment (dangerismycat) [inline docs/src/Module.doc.js:117]: Sadly this doesn't affect how the type is rendered, but it does make it a lot more readable here

File: packages/gestalt/src/Badge.js [hunk 5]
```diff
@@ -1,18 +1,19 @@
 // @flow strict
```
Comment (dangerismycat) [inline packages/gestalt/src/Badge.js:1]: only simple refactors/cleanup in this file, no material changes

File: packages/gestalt/src/Module.js [hunk 6]
```diff
@@ -4,14 +4,16 @@ import Box from './Box.js';
 import Flex from './Flex.js';
 import ModuleTitle from './ModuleTitle.js';
 import ModuleExpandable from './ModuleExpandable.js';
-import { type BaseModuleProps } from './moduleTypes.js';
+import { type BaseModuleTitleProps } from './moduleTypes.js';
 
 type Props = {|
-  ...BaseModuleProps,
+  ...BaseModuleTitleProps,
+  children?: Node,
```
Comment (dangerismycat) [inline packages/gestalt/src/Module.js:11]: this was previously in the common type

File: packages/gestalt/src/ModuleExpandable.js [hunk 7]
```diff
@@ -36,31 +34,34 @@ export default function ModuleExpandable({
...
-            summary={summary}
-            title={title}
-            type={type}
-          >
-            {children}
-          </ModuleExpandableItem>
+      {items.map(
+        ({ badgeText, children, icon, iconAccessibilityLabel, summary, title, type }, index) => (
+          <Fragment key={index}>
+            {index > 0 && <Divider />}
```
Comment (dangerismycat) [inline packages/gestalt/src/ModuleExpandable.js:40]: this is the same logic, just simpler (and more functional, no need to reference something outside the iterative loop)

File: packages/gestalt/src/ModuleTitle.js [hunk 8]
```diff
@@ -30,17 +31,29 @@ export default function ModuleTitle({
...
   return (
     <Flex gap={2}>
       {iconName && (
         <Icon accessibilityLabel={iconAccessibilityLabel ?? ''} color={color} icon={iconName} />
       )}
 
-      <Text color={color} truncate weight="bold">
-        {title}
-      </Text>
+      <Flex.Item minWidth={0}>
```
Comment (dangerismycat) [inline packages/gestalt/src/ModuleTitle.js:42]: this keeps the text from overflowing the container if very long

File: packages/gestalt/src/ModuleTitle.js [hunk 9]
```diff
@@ -30,17 +31,29 @@ export default function ModuleTitle({
...
+        <Text color={color} truncate weight="bold">
+          {title}
+        </Text>
+      </Flex.Item>
+
+      {badgeText && (
+        <Box
+          dangerouslySetInlineStyle={{ __style: { top: '1px' } }}
```
Comment (dangerismycat) [inline packages/gestalt/src/ModuleTitle.js:50]: nudged a hair for visual centering

Comment (AlbertCarreras) [reply packages/gestalt/src/ModuleTitle.js:50]: If you remove relative, do you still need the top px? I played with the css in the Docs and seemed to work, but might be wrong.

Comment (dangerismycat) [reply packages/gestalt/src/ModuleTitle.js:50]: Yep, the `position` is necessary for the `top` to take effect. And the `top` is needed to nudge the badge down a hair so it's visually vertically centered (as contrasted with what flex thinks vertical centering is)

Comment (AlbertCarreras) [reply packages/gestalt/src/ModuleTitle.js:50]: could the badge be centered on Flex? alignItems="center"? Just curious.

**Comment by AlbertCarreras** [comment]
LGTM!

**Review by AlbertCarreras — APPROVED** [review]

File: packages/gestalt/src/ModuleTitle.js (hunk 9 above)
Comment (dangerismycat) [reply packages/gestalt/src/ModuleTitle.js:50]: That's what I mean by "visually centered" as opposed to just "centered". This is a common occurrence when aligning text and other visual elements: what flex thinks is "centered" is actually off by a pixel or two.

**Pull request merged**